src/LowerTriangularMatrices/lower_triangular_matrix.jl

```diff
@@ -162,127 +239,130 @@ function Base.copyto!(  L1::LowerTriangularMatrix{T},   # copy to L1
     @inbounds for m in 1:mmax
         for l in m:lmax
             lm += 1
-            L1[lm] = (l in ls) && (m in ms) ? convert(T, L2[l, m]) : L1[lm]
+            L1[lm,[Colon() for i=1:(N-2)]...] = (l in ls) && (m in ms) ? T.(L2[l, m, [Colon() for i=1:(N-2)]...]) : L1[lm, [Colon() for i=1:(N-2)]...]
         end
     end
+
     L1
-end
+end 
+
+# Fallback / GPU version
+function Base.copyto!(  L1::LowerTriangularArray{T,N,ArrayType1},   # copy to L1
+                        L2::LowerTriangularArray{S,N,ArrayType2},      # copy from L2
+                        ls::AbstractUnitRange,          # range of indices in 1st dim
+                        ms::AbstractUnitRange) where {T, S, N, ArrayType1<:AbstractArray{T}, ArrayType2<:AbstractArray{S}}  # range of indices in 2nd dim
+
+    lmax, mmax = size(L2)        # use the size of L2 for boundscheck
+    @boundscheck maximum(ls) <= lmax || throw(BoundsError)
+    @boundscheck maximum(ms) <= mmax || throw(BoundsError)
+
+    # we'll setup the 1D-indices that correspond the given range here
+    ind_L1 = lowertriangle_indices(L1.m, L1.n)
+    ind_L1[ls[end]+1:end, :] .= 0 
+    ind_L1[:, ms[end+1]:end] .= 0 
+    ind_L1 = reshape(ind_L1, :)
+
+    ind_L2 = lowertriangle_indices(L2.m, L2.m) 
+    ind_L2[ls[end]+1:end, :] .= 0 
+    ind_L2[:, ms[end+1]:end] .= 0 
+    ind_L2 = reshape(ind_L2, :)
+
+    L1[ind_l1,[Colon() for i=1:(N-2)]...] = T.(L2[ind_l2,[Colon() for i=1:(N-2)]...])
 
-function Base.copyto!(  L::LowerTriangularMatrix{T},    # copy to L
+    L1
+end 
+
+function Base.copyto!(  L::LowerTriangularArray{T},    # copy to L
                         M::AbstractMatrix) where T      # copy from M
 
     @boundscheck size(L) == size(M) || throw(BoundsError)
-    lmax, mmax = size(L)
+    L.data .= convert.(T, M[lowertriangle_indices(M)])
 
-    lm = 0
-    @inbounds for m in 1:mmax
-        for l in m:lmax
-            lm += 1
-            L[lm] = convert(T, M[l, m])
-        end
-    end
     L
 end
 
-function Base.copyto!(  M::AbstractMatrix{T},               # copy to M
-                        L::LowerTriangularMatrix) where T   # copy from L
+function Base.copyto!(  M::AbstractArray{T},               # copy to M
+                        L::LowerTriangularArray) where T   # copy from L
 
     @boundscheck size(L) == size(M) || throw(BoundsError)
-    lmax, mmax = size(L)
 
-    lm = 0
-    @inbounds for m in 1:mmax
-        for l in 1:m-1          # zero for upper triangle (excl diagonal)
-            M[l, m] = zero(T)
-        end
+    lower_triangle_indices = lowertriangle_indices(M)
+    upper_triangle_indices = @. ~lower_triangle_indices
```

Comment:
When we talk about the "upper triangle" then we always mean excluding the diagonal, maybe worth adding this throughout as a comment to avoid confusion?